### DISCUSSION TITLE
Using pool_pre_ping but getting 'MySQL server has gone away'

### DISCUSSION THREAD

**Discussion by BackSeat** [sqlalchemy/sqlalchemy#6561] [Usage Questions]
I left an application running but idle overnight. When I tried to use it the following day, I got the 'MySQL server has gone away' message.

I have the engine configured like this:

```python
engine = sqlalchemy.create_engine(f"{Config.MYSQL_CONNECT}?charset=utf8",
                                  encoding='utf-8',
                                  echo=Config.DISPLAY_SQL,
                                  pool_pre_ping=True)
```

My understanding was that using `pool_pre_ping` would stop this error (assuming MySQL is actually still available, which it is). Have I misunderstood how this works?

MariaDB 10.5.10
SQLAlchemy 1.4.2

**Comment by CaselIT** [comment]
Hi,

This is only for newly created connection, gotten from the engine. Meaning that if you have a connection checked out, `pool_pre_ping` would not have any effect.

Do you receive this error on a newly obtained connection or when re-using an already checked out one?

**Reply by BackSeat** [reply]
I don't understand what you mean by a "checked out" connection.

The code I'm using to create a session is below. Is that wrong?

```python
engine = sqlalchemy.create_engine(f"{Config.MYSQL_CONNECT}?charset=utf8",
                                  encoding='utf-8',
                                  echo=Config.DISPLAY_SQL,
                                  pool_pre_ping=True)
Base = declarative_base()
Base.metadata.create_all(engine)

Session = sessionmaker(bind=engine)
session = Session()
```

**Reply by CaselIT** [reply]
> I don't understand what you mean by a "checked out" connection.

I was referring to the connection pool used by sqlalchemy. Checked out refers to a connection that is in use and not returned to the pool. See also https://docs.sqlalchemy.org/en/14/core/pooling.html

In your example, if you never close the session, or create a new one, it may keep a connection checked out from the sqlalchemy pool until it's gc by python.

**Reply by zzzeek** [reply]
please specify the MySQL DBAPI you are using, some have different behaviors regarding ping, for example mysql-python's ping feature handles the reconnect internally.

**Reply by BackSeat** [reply]
[mysqlclient](https://docs.sqlalchemy.org/en/14/dialects/mysql.html#module-sqlalchemy.dialects.mysql.mysqldb), but happy to change if there's a more suitable DBAPI for my use case.

**Reply by zzzeek** [reply]
nope, I just gave it a test here, worked great, i had echo_pool=True on to see what it was doing and just had noticed that the pre-ping seemed to return success even though I had manually killed the connection, this is apparently due to MySQL auto-reconnect: https://dev.mysql.com/doc/c-api/5.7/en/mysql-ping.html

set echo_pool=True on create_engine() , put your application into the same idle state that you left it with overnight, then manually log in to MySQL , do SHOW PROCESSLIST then "KILL" the connection that is for your application.  then resume your app as you did before and that will indicate the state transition.  if you dont see any "pool checkout " / "pre-ping" logging messages that means your app's idle state has the connecvton checked out.

**Reply by BackSeat** [reply]
I see no messages up to and including killing the MySQL connection. When I then try to use the app, I get

```
2021-05-31 19:08:25,884 INFO sqlalchemy.pool.impl.QueuePool Invalidate connection <_mysql.connection open to 'localhost' at 0x1baf3c0> (reason: OperationalError:(2006, 'MySQL server has gone away'))
```

followed by an exception in the app: 
`MySQLdb._exceptions.OperationalError) (2006, 'MySQL server has gone away')`

**Reply by CaselIT** [reply]
From that it seems that you have a connection that is checked out during the idle period.
As mentioned above, pre_ping only works for newly created connections or session (or at least the session should be closed)

In the documentation you find various suggestion on how to use context managers to handle the session life, like https://docs.sqlalchemy.org/en/14/orm/session_basics.html#session-faq-whentocreate

**Reply by BackSeat** [reply]
@CaselIT @zzzeek Thanks for your help. I had misunderstood how Session should be used. On the right track now.
